### PR TITLE
Add BARTpho: Pre-trained Sequence-to-Sequence Models for Vietnamese

### PR DESCRIPTION
**Pull request by datquocnguyen**
# What does this PR do?

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Please help have a look: @patrickvonplaten, @patil-suraj Thanks.

### CODE REVIEW

File: docs/source/model_doc/bartpho.rst [hunk 1]
```diff
@@ -0,0 +1,64 @@
+..
+    Copyright 2020 The HuggingFace Team. All rights reserved.
```
Comment (patil-suraj) [inline docs/source/model_doc/bartpho.rst:2]:
```suggestion
    Copyright 2021 The HuggingFace Team. All rights reserved.
```

File: docs/source/model_doc/bartpho.rst [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+research and applications of generative Vietnamese NLP tasks.*
+
+Example of use:
+
+.. code-block::
+
+    >>> import torch
+    >>> from transformers import AutoModel, AutoTokenizer
+
+    >>> bartpho = AutoModel.from_pretrained("vinai/bartpho-syllable")
```
Comment (patil-suraj) [inline docs/source/model_doc/bartpho.rst:38]: This is a seq2seq model, so we should use `AutoModelForSeq2SeqLM` instead of `AutoModel`.

File: docs/source/model_doc/bartpho.rst [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    >>> line = "Chúng tôi là những nghiên cứu viên."
+
+    >>> input_ids = torch.tensor([tokenizer.encode(line)])
+
+    >>> with torch.no_grad():
+    ...     features = bartpho(input_ids)  # Models outputs are now tuples
+
+    >>> # With TensorFlow 2.0+:
+    >>> # from transformers import TFAutoModel
+    >>> # bartpho = TFAutoModel.from_pretrained("vinai/bartpho-syllable")
```
Comment (patil-suraj) [inline docs/source/model_doc/bartpho.rst:55]: same comment as above

Comment (patil-suraj) [reply docs/source/model_doc/bartpho.rst:55]: also, this shouldn't be commented

File: docs/source/model_doc/bertweet.rst [hunk 4]
```diff
@@ -10,7 +10,7 @@
     an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
     specific language governing permissions and limitations under the License.
 
-Bertweet
+BERTweet
```
Comment (patil-suraj) [inline docs/source/model_doc/bertweet.rst:13]: This should be done in separate PR

File: docs/source/model_doc/phobert.rst [hunk 5]
```diff
@@ -50,7 +50,8 @@ Example of use:
     >>> # phobert = TFAutoModel.from_pretrained("vinai/phobert-base")
 
 
-    This model was contributed by `dqnguyen <https://huggingface.co/dqnguyen>`__. The original code can be found `here <https://github.com/VinAIResearch/PhoBERT>`__.
+This model was contributed by `dqnguyen <https://huggingface.co/dqnguyen>`__. The original code can be found `here
+<https://github.com/VinAIResearch/PhoBERT>`__.
```
Comment (patil-suraj) [inline docs/source/model_doc/phobert.rst:54]: same comment as above

File: src/transformers/models/bartpho/__init__.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
+# flake8: noqa
+# There's no way to ignore "F401 '...' imported but unused" warnings in this
+# module, but to preserve other warnings. So, don't check this module at all.
+
+# Copyright 2020 The HuggingFace Team. All rights reserved.
```
Comment (patil-suraj) [inline src/transformers/models/bartpho/__init__.py:5]:
```suggestion
# Copyright 2021 The HuggingFace Team. All rights reserved.
```

File: src/transformers/models/bartpho/tokenization_bartpho.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+            sep_token=sep_token,
+            cls_token=cls_token,
+            pad_token=pad_token,
+            mask_token=mask_token,
+            sp_model_kwargs=self.sp_model_kwargs,
+            **kwargs,
+        )
+
+        self.vocab_file = vocab_file
+        self.reduced_vocab_file = reduced_vocab_file
```
Comment (patil-suraj) [inline src/transformers/models/bartpho/tokenization_bartpho.py:147]: what is `reduced_vocab_file` ? From a quick glance, it seems `sentencepiece` is only used to tokenize the text, and the reduced_vocab_file is used to convert tokens to ids and ids to tokens. Is that right? And if it is so, could we call 
sentencepiece file as `spm_file` and `reduced_vocab_file` as the `vocab_file` since `sentencepiece` is not really used as vocab here.


 [`M2M100Tokenizer`](https://github.com/huggingface/transformers/blob/master/src/transformers/models/m2m_100/tokenization_m2m_100.py) is also similar to this, where it uses `sentencepiece` just for tokenization.

File: src/transformers/models/bartpho/tokenization_bartpho.py [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+
+from ...tokenization_utils import AddedToken, PreTrainedTokenizer
+from ...utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+SPIECE_UNDERLINE = "▁"
+
+VOCAB_FILES_NAMES = {"vocab_file": "sentencepiece.bpe.model", "reduced_vocab_file": "dict.txt"}
```
Comment (patil-suraj) [inline src/transformers/models/bartpho/tokenization_bartpho.py:32]: The vocab file name for  [bartpho-word](https://huggingface.co/vinai/bartpho-word/tree/main) is `vocab.txt`, it should also be changed to `dict.txt` otherwise it' 'll fail to load, since the file names need to match with specified file names here.

File: src/transformers/models/bartpho/tokenization_bartpho.py [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+
+        self.sp_model = spm.SentencePieceProcessor(**self.sp_model_kwargs)
+        self.sp_model.LoadFromSerializedProto(self.sp_model_proto)
+
+    def build_inputs_with_special_tokens(
+        self, token_ids_0: List[int], token_ids_1: Optional[List[int]] = None
+    ) -> List[int]:
+        """
+        Build model inputs from a sequence or a pair of sequence for sequence classification tasks by concatenating and
+        adding special tokens. An XLM-RoBERTa sequence has the following format:
```
Comment (patil-suraj) [inline src/transformers/models/bartpho/tokenization_bartpho.py:182]:
```suggestion
        adding special tokens. An BARTPho sequence has the following format:
```

File: src/transformers/models/bartpho/tokenization_bartpho.py [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+
+        if token_ids_1 is None:
+            return [1] + ([0] * len(token_ids_0)) + [1]
+        return [1] + ([0] * len(token_ids_0)) + [1, 1] + ([0] * len(token_ids_1)) + [1]
+
+    def create_token_type_ids_from_sequences(
+        self, token_ids_0: List[int], token_ids_1: Optional[List[int]] = None
+    ) -> List[int]:
+        """
+        Create a mask from the two sequences passed to be used in a sequence-pair classification task. XLM-RoBERTa does
```
Comment (patil-suraj) [inline src/transformers/models/bartpho/tokenization_bartpho.py:235]:
```suggestion
        Create a mask from the two sequences passed to be used in a sequence-pair classification task.BARTPho does
```

File: src/transformers/models/bartpho/__init__.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+_import_structure = {
+    "tokenization_bartpho": ["BartphoTokenizer"],
+}
+
+
+if TYPE_CHECKING:
+    from .tokenization_bartpho import BartphoTokenizer
+
+else:
+    import sys
```
Comment (patil-suraj) [inline src/transformers/models/bartpho/__init__.py:33]:
```suggestion
from ...file_utils import _LazyModule, is_sentencepiece_available


_import_structure = {}

if is_sentencepiece_available():
    _import_structure["tokenization_bartpho"] = ["BartphoTokenizer"]

if TYPE_CHECKING:
    if is_sentencepiece_available():
        from .tokenization_bartpho import BartphoTokenizer

else:
    import sys
```

This tokenizer depends on `sentencepiece` so it should be imported only if `sentencepiece` is installed.

File: tests/test_tokenization_bartpho.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
+# coding=utf-8
+# Copyright 2020 HuggingFace Inc. team.
```
Comment (patil-suraj) [inline tests/test_tokenization_bartpho.py:2]:
```suggestion
# Copyright 2021 HuggingFace Inc. team.
```

File: tests/test_tokenization_bartpho.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+from .test_tokenization_common import TokenizerTesterMixin
+
+
+SAMPLE_VOCAB = os.path.join(os.path.dirname(os.path.abspath(__file__)), "fixtures/test_sentencepiece_bpe.model")
+
+
+class BartphoTokenizerTest(TokenizerTesterMixin, unittest.TestCase):
+
+    tokenizer_class = BartphoTokenizer
+    test_rust_tokenizer = False
```
Comment (patil-suraj) [inline tests/test_tokenization_bartpho.py:30]:
```suggestion
    test_rust_tokenizer = False
    test_sentencepiece = True
```

File: tests/test_tokenization_bartpho.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+    def test_full_tokenizer(self):
+        tokenizer = BartphoTokenizer(SAMPLE_VOCAB, self.reduced_vocab_file, **self.special_tokens_map)
+        text = "This is a là test"
+        bpe_tokens = "▁This ▁is ▁a ▁l à ▁t est".split()
+        tokens = tokenizer.tokenize(text)
+        self.assertListEqual(tokens, bpe_tokens)
+
+        input_tokens = tokens + [tokenizer.unk_token]
+        input_bpe_tokens = [4, 5, 6, 3, 3, 7, 8, 3]
```
Comment (patil-suraj) [inline tests/test_tokenization_bartpho.py:64]: nice test! could we also add an integration test the main tokenizer, An example could found here

https://github.com/huggingface/transformers/blob/3a8a8013adc5802cc302e63e403f0e8925fab0ea/tests/test_tokenization_m2m_100.py#L112-L122

File: docs/source/model_doc/bartpho.rst [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    >>> import torch
+    >>> from transformers import AutoModel, AutoTokenizer
+
+    >>> bartpho = AutoModel.from_pretrained("vinai/bartpho-syllable")
+
+    >>> # For transformers v4.x+:
+    >>> tokenizer = AutoTokenizer.from_pretrained("vinai/bartpho-syllable", use_fast=False)
+
+    >>> # For transformers v3.x:
+    >>> # tokenizer = AutoTokenizer.from_pretrained("vinai/bartpho-syllable")
```
Comment (patil-suraj) [inline docs/source/model_doc/bartpho.rst:44]: if a fast tokenizer is not available it'll automatically fall back to slow tokenizer, so there is no need to pass `use_fast=False`.

**Review by patil-suraj — COMMENTED** [review]
Thanks a lot for adding this, great to have `BARTpho` integrated into Transformers!

I have left a few comments below, let me know if something is not clear, happy to help :)

Also AFAIK it's not possible to add a fast version of this type of tokenizer where `sentencepiece` is used only for tokenization and a different vocab file is used to id to token and token to id conversion. But @SaulLu , @n1t0  would know more :) 
(cc @LysandreJik @sgugger)

File: docs/source/index.rst [hunk 16]
```diff
@@ -553,6 +553,7 @@ Flax), PyTorch, and/or TensorFlow.
     model_doc/auto
     model_doc/bart
     model_doc/barthez
+    model_doc/bartpho
```
Comment (sgugger) [inline docs/source/index.rst:556]: Don't forget to add your model to the main README (in alphabetical order) then run `make fix-copies` to update this file as well.

File: docs/source/model_doc/bartpho.rst (hunk 15 above)
Comment (sgugger) [inline docs/source/model_doc/bartpho.rst:44]:
```suggestion
    >>> # tokenizer = AutoTokenizer.from_pretrained("vinai/bartpho-syllable")
```
Will work for any version.

File: docs/source/model_doc/bartpho.rst [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+
+    >>> line = "Chúng tôi là những nghiên cứu viên."
+
+    >>> input_ids = torch.tensor([tokenizer.encode(line)])
```
Comment (sgugger) [inline docs/source/model_doc/bartpho.rst:48]:
```suggestion
    >>> input_ids = tokenizer(line, return_tensors="pt)["input_ids"]
```
We prefer to advertise the call method in the doc.

File: src/transformers/models/bartpho/tokenization_bartpho.py [hunk 18]
```diff
@@ -0,0 +1,302 @@
...
+    <https://github.com/google/sentencepiece>`__.
+
+    This tokenizer inherits from :class:`~transformers.PreTrainedTokenizer` which contains most of the main methods.
+    Users should refer to this superclass for more information regarding those methods.
+
+    Args:
+        vocab_file (:obj:`str`):
+            Path to the vocabulary file.
+        reduced_vocab_file (:obj:`str`):
+            Path to the reduced vocabulary file.
```
Comment (sgugger) [inline src/transformers/models/bartpho/tokenization_bartpho.py:58]: Could you elaborate on the difference with the `vocab_file`?

**Review by sgugger — COMMENTED** [review]
I wonder how this tokenizer exactly differs from a classic sentencepiece tokenizer (like BART)? Could we also add the fast version?

File: docs/source/model_doc/bartpho.rst (hunk 2 above)
Comment (datquocnguyen) [reply docs/source/model_doc/bartpho.rst:38]: The config file `config.json` points to BART seq2seq model. 
So it would be totally fine using AutoModel, wouldn't it?

File: docs/source/model_doc/bertweet.rst (hunk 4 above)
Comment (datquocnguyen) [reply docs/source/model_doc/bertweet.rst:13]: BERTweet & PhoBERT are also my models, that why I made some changes to them.

File: src/transformers/models/bartpho/tokenization_bartpho.py (hunk 7 above)
Comment (datquocnguyen) [reply src/transformers/models/bartpho/tokenization_bartpho.py:147]: I did not train a sentencepiece for Vietnamese. 
BARTpho-syllable employs the existing pre-trained "sentencepiece" model from XLMRoBERTaTokenizer, and this pre-trained model is referred to as a vocab of 250K types. 
`reduced_vocab_file` is a vocab containing 40 Vietnamese-focused types extracted from the XLMRoBERTaTokenizer vocab of 250K types. 
Other languages can thus simply reuse BartphoTokenizer with their own `reduced_vocab_file`.

File: src/transformers/models/bartpho/tokenization_bartpho.py (hunk 8 above)
Comment (datquocnguyen) [reply src/transformers/models/bartpho/tokenization_bartpho.py:32]: No, it will not. BartphoTokenizer is used for `bartpho-syllable` only. 

`barpho-word` employs PhobertTokenizer, as shown in its `config.json` file and our paper: https://arxiv.org/pdf/2109.09701.pdf

The difference between `bartpho-syllable` and `bartpho-word` is that they take different types of input texts: syllable level for `bartpho-syllable` (e.g. 7-syllable written text “chúng tôi là những nghiên cứu viên”) vs. word level for `barpho-word` (e.g. 4-word text “chúng_tôi là những nghiên_cứu_viên”). So they use different tokenizers.

File: tests/test_tokenization_bartpho.py (hunk 14 above)
Comment (datquocnguyen) [reply tests/test_tokenization_bartpho.py:64]: Is this necessary? I am not sure.
This M2M100Tokenizer has an opposite operating manner compared to ours. M2M100Tokenizer allows adding extra types from new languages when continuing pre-training. 
Ours, on the other hand, wants to reduce the vocab size of an existing one (e.g. equivalent to reducing model sizes of XLM-RoBERTa or mBART when working with only Vietnamese).

File: docs/source/model_doc/bartpho.rst (hunk 15 above)
Comment (datquocnguyen) [reply docs/source/model_doc/bartpho.rst:44]: Thanks. I will revise and commit.

File: src/transformers/models/bartpho/__init__.py (hunk 11 above)
Comment (datquocnguyen) [reply src/transformers/models/bartpho/__init__.py:33]: Thanks. I will revise and commit.

File: docs/source/index.rst (hunk 16 above)
Comment (datquocnguyen) [reply docs/source/index.rst:556]: I tried adding my other two models BERTweet and PhoBERT to the main README when making first commits 1+ years ago, but got rejections. Reason: the only thing I contributed is the new language/domain-specific tokenizers (not modeling). 
Can I now add BARTpho as well as BERTweet (EMNLP 2020 Demonstrations) and PhoBERT (EMNLP 2020 Findings) into the main README file?

File: src/transformers/models/bartpho/tokenization_bartpho.py (hunk 18 above)
Comment (datquocnguyen) [reply src/transformers/models/bartpho/tokenization_bartpho.py:58]: I did not train a sentencepiece for Vietnamese.
BARTpho-syllable employs the existing pre-trained "sentencepiece" model from XLMRoBERTaTokenizer, and this pre-trained model is referred to as a vocab of 250K types.
reduced_vocab_file is a vocab containing 40 Vietnamese-focused types extracted from the XLMRoBERTaTokenizer vocab of 250K types.
Other languages can thus simply reuse BartphoTokenizer with their own reduced_vocab_file.

**Comment by datquocnguyen** [comment]
Thanks, @patil-suraj and @sgugger for helpful feedback.
The same comment I get from both of you is regarding the vocab_file. Here is a summary:

I did not train a sentencepiece for Vietnamese.
bartpho-syllable employs the existing pre-trained "sentencepiece" model from XLMRoBERTaTokenizer, and this pre-trained model is referred to as a `vocab_file` of 250K types.
`reduced_vocab_file` is a vocab containing 40 Vietnamese-focused types extracted from the XLMRoBERTaTokenizer vocab of 250K types.

Usecase of BartphoTokenizer: Other languages can thus simply reuse BartphoTokenizer with their own `reduced_vocab_file`. The goal here is to reduce model sizes of existing pre-trained XLM-RoBERTa/mBART models when applying to a small set of languages instead of the whole 50/100 languages.

File: docs/source/model_doc/phobert.rst (hunk 5 above)
Comment (datquocnguyen) [reply docs/source/model_doc/phobert.rst:54]: BERTweet & PhoBERT are also my models, that why I made some changes to them.

File: src/transformers/models/bartpho/tokenization_bartpho.py (hunk 18 above)
Comment (patrickvonplaten) [reply src/transformers/models/bartpho/tokenization_bartpho.py:58]: It would be great if you could add 1 or 2 sentences directly to the docstring :-)

File: src/transformers/models/bartpho/tokenization_bartpho.py [hunk 19]
```diff
@@ -0,0 +1,302 @@
...
+            :obj:`List[int]`: List of zeros.
+
+        """
+
+        sep = [self.sep_token_id]
+        cls = [self.cls_token_id]
+
+        if token_ids_1 is None:
+            return len(cls + token_ids_0 + sep) * [0]
+        return len(cls + token_ids_0 + sep + sep + token_ids_1 + sep) * [0]
```
Comment (patrickvonplaten) [inline src/transformers/models/bartpho/tokenization_bartpho.py:254]: this looks like BERT or RoBERTa - I don't think BART makes use of the CLS token no? 
Think it would be better to write the function more like it is in T5: https://github.com/huggingface/transformers/blob/3a8de58c5192b620228128430ea52e6eda81c40a/src/transformers/models/t5/tokenization_t5.py#L217

Comment (patrickvonplaten) [reply src/transformers/models/bartpho/tokenization_bartpho.py:254]: With the following statement: https://github.com/huggingface/transformers/blob/3a8de58c5192b620228128430ea52e6eda81c40a/src/transformers/models/t5/tokenization_t5.py#L201

Comment (patrickvonplaten) [reply src/transformers/models/bartpho/tokenization_bartpho.py:254]: Bart is a bad example as it inherits from RoBERTa

File: src/transformers/models/bartpho/tokenization_bartpho.py [hunk 20]
```diff
@@ -0,0 +1,302 @@
...
+        # Mask token behave like a normal word, i.e. include the space before it
+        mask_token = AddedToken(mask_token, lstrip=True, rstrip=False) if isinstance(mask_token, str) else mask_token
+
+        self.sp_model_kwargs = {} if sp_model_kwargs is None else sp_model_kwargs
+
+        super().__init__(
+            bos_token=bos_token,
+            eos_token=eos_token,
+            unk_token=unk_token,
+            sep_token=sep_token,
```
Comment (patrickvonplaten) [inline src/transformers/models/bartpho/tokenization_bartpho.py:138]: I don't think those tokens are used no? `sep_token` and `cls_token` - do we need them for anything?

File: docs/source/model_doc/bartpho.rst [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+
+    >>> with torch.no_grad():
+    ...     features = bartpho(**input_ids)  # Models outputs are now tuples
+
+    >>> # With TensorFlow 2.0+:
+    >>> from transformers import TFAutoModel
+    >>> bartpho = TFAutoModel.from_pretrained("vinai/bartpho-syllable")
+    >>> input_ids = tokenizer(line, return_tensors="tf")
+    >>> features = bartpho(**input_ids)
+
```
Comment (SaulLu) [inline docs/source/model_doc/bartpho.rst:54]: Perhaps this model would benefit from having a "Tips" section explaining how it differs from `MBart` and `XLMRoberta` what do you think? :slightly_smiling_face: 

I have in mind BertJapanese's [overview section](https://huggingface.co/transformers/model_doc/bert_japanese.html#overview) for example.

How about also using this `Tips` section to explain how to use this architecture to train a model with a new `reduced_vocab_file` (if I am not mistaken, this particularity is not explained anywhere else)? (cc @sgugger , @LysandreJik, @patil-suraj and @patrickvonplaten )?

File: src/transformers/models/bartpho/tokenization_bartpho.py (hunk 18 above)
Comment (SaulLu) [reply src/transformers/models/bartpho/tokenization_bartpho.py:58]: I join the previous comments, I think it would be really nice to have some more details about `reduced_vocab_file` in this docstring.

**Review by SaulLu — COMMENTED** [review]
Thank you so much for this addition and your work @datquocnguyen !

Concerning tokenizer fast, indeed I don't think that our tokenizer fast were designed to support this kind of behavior. It might be possible to find a trick to make it work but unfortunately I don't see it now.

**Review by LysandreJik — APPROVED** [review]
This looks good to me @datquocnguyen, thank you for adding another tokenizer alongside BERTweet and PhoBERT!

Too bad that there can be no fast tokenizer, but this LGTM either way!

**Comment by datquocnguyen** [comment]
Thanks @LysandreJik 

My pull request suddenly failed the check `run_tests_torch_and_flax`:

`FAILED tests/test_modeling_flax_clip.py::FlaxCLIPModelTest::test_equivalence_flax_to_pt`
`FAILED tests/test_modeling_flax_clip.py::FlaxCLIPModelTest::test_equivalence_pt_to_flax`

They are out of my control, not relating to BartphoTokenizer.

**Review by sgugger — APPROVED** [review]
One last small comment and you should then run `make fix-copies`. Then we will be good to merge :-)
